### PR TITLE
rt refine: progress on Finalise_R sorries

### PR DESCRIPTION
**Pull request by corlewis**

### CODE REVIEW

File: proof/refine/ARM/Finalise_R.thy [hunk 1]
```diff
@@ -2256,53 +2256,90 @@ lemma ntfn_bound_tcb_at':
...
-   apply (frule ko_at_state_refs_ofD', simp)
-  apply (clarsimp split: if_split_asm)
-   apply (frule_tac P="(=) (Some ntfnptr)" in ntfn_bound_tcb_at', simp_all add: valid_pspace'_def)[1]
-   subgoal by (fastforce simp: ntfn_q_refs_of'_def state_refs_of'_def tcb_ntfn_is_bound'_def
-                          tcb_st_refs_of'_def
-                   dest!: bound_tcb_at_state_refs_ofD'
-                   split: ntfn.splits thread_state.splits)
-  apply (frule ko_at_state_refs_ofD', simp)
-  done *)
+  apply (wpsimp wp: sbn'_valid_pspace'_inv sbn_sch_act' sbn_valid_queues
```
Comment (michaelmcinerney) [inline proof/refine/ARM/Finalise_R.thy:2263]: Nitpick: indentation. And is the `+` required?

File: proof/refine/ARM/Finalise_R.thy [hunk 2]
```diff
@@ -2256,53 +2256,90 @@ lemma ntfn_bound_tcb_at':
...
+  done
+
+lemma setSchedContext_utr[wp]:
+  "\<lbrace>untyped_ranges_zero'\<rbrace> setSchedContext scPtr sc \<lbrace>\<lambda>rv. untyped_ranges_zero'\<rbrace>"
+  apply (simp add: cteCaps_of_def)
+  apply (rule hoare_pre, wp untyped_ranges_zero_lift)
+  apply (simp add: o_def)
+  done
+
+lemma schedContextUnbindNtfn_invs'[wp]:
```
Comment (michaelmcinerney) [inline proof/refine/ARM/Finalise_R.thy:2296]: We'll need `schedContextUnbindNtfn_valid_objs'` in Ipc_R. Would you be able to factor that out into a separate lemma and use it in this proof?

File: proof/refine/ARM/Finalise_R.thy [hunk 3]
```diff
@@ -2256,53 +2256,90 @@ lemma ntfn_bound_tcb_at':
...
+  "\<lbrace>invs'\<rbrace> schedContextUnbindNtfn scPtr \<lbrace>\<lambda>rv. invs'\<rbrace>"
+  unfolding schedContextUnbindNtfn_def invs'_def valid_state'_def
+  apply (wpsimp wp: getNotification_wp hoare_vcg_all_lift hoare_vcg_imp_lift'
+                    typ_at_lifts valid_ntfn_lift' valid_pde_mappings_lift')
+  apply (normalise_obj_at')
```
Comment (michaelmcinerney) [inline proof/refine/ARM/Finalise_R.thy:2301]: Brackets around `normalise_obj_at'` aren't necessary

File: proof/refine/ARM/Finalise_R.thy [hunk 4]
```diff
@@ -2256,53 +2256,90 @@ lemma ntfn_bound_tcb_at':
...
+
+lemma setSchedContext_utr[wp]:
+  "\<lbrace>untyped_ranges_zero'\<rbrace> setSchedContext scPtr sc \<lbrace>\<lambda>rv. untyped_ranges_zero'\<rbrace>"
+  apply (simp add: cteCaps_of_def)
+  apply (rule hoare_pre, wp untyped_ranges_zero_lift)
+  apply (simp add: o_def)
+  done
+
+lemma schedContextUnbindNtfn_invs'[wp]:
+  "\<lbrace>invs'\<rbrace> schedContextUnbindNtfn scPtr \<lbrace>\<lambda>rv. invs'\<rbrace>"
```
Comment (michaelmcinerney) [inline proof/refine/ARM/Finalise_R.thy:2297]: Maybe use invariant notation?

File: proof/refine/ARM/Finalise_R.thy [hunk 5]
```diff
@@ -2723,6 +2771,87 @@ crunch invs[wp]: prepareThreadDelete "invs'"
...
+  apply (frule_tac p=scPtr in sym_refs_obj_atD', assumption)
+  apply (clarsimp simp: ko_wp_at'_def obj_at'_def projectKOs get_refs_def2)
+  done
+
+lemma schedContextUnbindNtfn_obj_at'_ntfnSc:
+  "\<lbrace>valid_objs' and obj_at' (\<lambda>ntfn. ntfnSc ntfn = Some scPtr) ntfnPtr\<rbrace>
+   schedContextUnbindNtfn scPtr
+   \<lbrace>\<lambda>_ s. obj_at' (\<lambda>ntfn. ntfnSc ntfn = None) ntfnPtr s\<rbrace>"
+  apply (simp add: schedContextUnbindNtfn_def sym_refs_asrt_def)
+  apply (rule hoare_seq_ext[OF _ stateAssert_sp])
```
Comment (michaelmcinerney) [inline proof/refine/ARM/Finalise_R.thy:2792]: Maybe roll this line into a `stateAssert_wp` in the following block?

**Review by michaelmcinerney — APPROVED** [review]
Cool! Good to see so many lemmas for new functions!

File: proof/refine/ARM/Finalise_R.thy (hunk 2 above)
Comment (corlewis) [reply proof/refine/ARM/Finalise_R.thy]: Done. Thanks for the suggestion, it even ended up improving the performance of the proof as well!

File: proof/refine/ARM/Finalise_R.thy [hunk 6]
```diff
@@ -2723,6 +2779,86 @@ crunch invs[wp]: prepareThreadDelete "invs'"
...
+  apply (wpsimp wp: hoare_vcg_all_lift threadGet_wp simp: inQ_def cong: conj_cong)
+  apply (clarsimp simp: obj_at'_def)
+  done
+
+lemma schedContextUnbindTCB_invs':
+  "\<lbrace>invs' and sch_act_simple and K (scPtr \<noteq> idle_sc_ptr)\<rbrace>
+   schedContextUnbindTCB scPtr
+   \<lbrace>\<lambda>_. invs'\<rbrace>"
+  unfolding schedContextUnbindTCB_def
+  sorry
```
Comment (mktnk3) [inline proof/refine/ARM/Finalise_R.thy:2860]: Can you please either sort this out before merging or label it with FIXME RT + some comments?

**Review by mktnk3 — APPROVED** [review]
Nice and neat progress :)

**Comment by corlewis** [comment]
Ok, I think this is now ready to be merged. There were some minor changes to a couple of `valid_queues` lemmas after rebasing, and otherwise the main difference from the previous version is that I removed the `replyRemove` lemmas while I continue working on them separately.

**Review by MitchellBuckley — APPROVED** [review]
Very nice! I have no comments to make except: 
* does this increase sorry-count, and are we ok with that?

**Comment by corlewis** [comment]
>     * does this increase sorry-count, and are we ok with that?

It does, and I can't speak for others but I'm ok with it. The alternative is I keep that part local or on a branch, and that would just make it more confusing for me and Victor.

**Comment by lsf37** [comment]
> It does, and I can't speak for others but I'm ok with it. The alternative is I keep that part local or on a branch, and that would just make it more confusing for me and Victor.

There is no problem with increasing the sorry count, it doesn't necessarily always go down.

**Comment by mktnk3** [comment]
I would think this is ok to go in, so long as things are clear between Corey and Victor with how to proceed. It is only one increase (or two? It’s a bit harder to search on GitHub) and I expect that it will be resolved fairly soon anyway.

**Review by VictorPhan — APPROVED** [review]
Looks good to me! Looking forward to working on these sorries.

**Pull request merged**